File: server/go/continuous_axis/continuous_axis_test.go
/*
	Copyright 2023 Google Inc.
	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at
		https://www.apache.org/licenses/LICENSE-2.0
	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	See the License for the specific language governing permissions and
	limitations under the License.
*/

package continuousaxis

import (
	"testing"
	"time"

	"github.com/google/traceviz/server/go/category"
	testutil "github.com/google/traceviz/server/go/test_util"
	"github.com/google/traceviz/server/go/util"
)

const timeLayout = "Jan 2, 2006 at 3:04pm (MST)"

func TestAxis(t *testing.T) {
	refTime, err := time.Parse(timeLayout, "Jan 1, 2020 at 1:00am (PST)")
	if err != nil {
		t.Fatalf("failed to parse reference time: %s", err)
	}
	ts := func(offset time.Duration) time.Time {
		return refTime.Add(offset)
	}
	cat := category.New("axis", "My axis", "All about my axis")
	for _, test := range []struct {
		description string
		axis        Axis
		wantUpdates []util.PropertyUpdate
		wantValues  map[interface{}]util.PropertyUpdate
	}{{
		description: "timestamp",
		axis:        NewTimestampAxis(cat, ts(0), ts(100)),
		wantUpdates: []util.PropertyUpdate{
			cat.Define(),
			util.StringProperty(axisTypeKey, timestampAxisType),
			util.TimestampProperty(axisMinKey, ts(0)),
			util.TimestampProperty(axisMaxKey, ts(100)),
		},
		wantValues: map[interface{}]util.PropertyUpdate{
			ts(10): util.TimestampProperty("axis", ts(10)),
		},
	}, {
		description: "duration",
		axis:        NewDurationAxis(cat, 100*time.Second),
		wantUpdates: []util.PropertyUpdate{
			cat.Define(),
			util.StringProperty(axisTypeKey, durationAxisType),
			util.DurationProperty(axisMinKey, 0),
			util.DurationProperty(axisMaxKey, 100*time.Second),
		},
		wantValues: map[interface{}]util.PropertyUpdate{
			10 * time.Second: util.DurationProperty("axis", 10*time.Second),
		},
	}, {
		description: "double",
		axis:        NewDoubleAxis(cat, 0, 100),
		wantUpdates: []util.PropertyUpdate{
			cat.Define(),
			util.StringProperty(axisTypeKey, doubleAxisType),
			util.DoubleProperty(axisMinKey, 0),
			util.DoubleProperty(axisMaxKey, 100),
		},
		wantValues: map[interface{}]util.PropertyUpdate{
			5.5: util.DoubleProperty("axis", 5.5),
		},
	}} {
		t.Run(test.description, func(t *testing.T) {
			gotUpdates := test.axis.Define()
			if msg, failed := testutil.NewUpdateComparator().
				WithTestUpdates(gotUpdates).
				WithWantUpdates(test.wantUpdates...).
				Compare(t); failed {
				t.Fatal(msg)
			}
			gotValues := map[interface{}]util.PropertyUpdate{}
			for val := range test.wantValues {
				gotValues[val] = test.axis.Value(val)
			}
			for val := range test.wantValues {
				if msg, failed := testutil.NewUpdateComparator().
					WithTestUpdates(test.wantValues[val]).
					WithWantUpdates(gotValues[val]).
					Compare(t); failed {
					t.Fatalf("Unexpected value for '%v': %s", val, msg)
				}
			}
		})
	}
}

File: client/core/src/continuous_axis/continuous_axis_test.ts
/*
        Copyright 2023 Google Inc.
        Licensed under the Apache License, Version 2.0 (the "License");
        you may not use this file except in compliance with the License.
        You may obtain a copy of the License at
                https://www.apache.org/licenses/LICENSE-2.0
        Unless required by applicable law or agreed to in writing, software
        distributed under the License is distributed on an "AS IS" BASIS,
        WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
        See the License for the specific language governing permissions and
        limitations under the License.
*/

import 'jasmine';

import { DurationAxis, getAxis, NumberAxis, TimestampAxis, unionAxes } from './continuous_axis.js';
import { Duration } from '../duration/duration.js';
import { Timestamp } from '../timestamp/timestamp.js';
import { dbl, dur, str, ts, valueMap } from '../value/test_value.js';

function sec(sec: number): Timestamp {
  return new Timestamp(sec, 0);
}

function d(nanos: number): Duration {
  return new Duration(nanos);
}

describe('axis test', () => {
  it('defines timestamp axes and gets points', () => {
    const axis = getAxis(valueMap(
        {key: 'category_defined_id', val: str('x_axis')},
        {key: 'category_display_name', val: str('time from start')},
        {key: 'category_description', val: str('Time from start')},
        {key: 'axis_type', val: str('timestamp')},
        {key: 'axis_min', val: ts(sec(0))},
        {key: 'axis_max', val: ts(sec(100))},
        ));
    expect(axis).toBeInstanceOf(TimestampAxis);
    const tsAxis = axis as TimestampAxis;
    expect(tsAxis.contains(sec(50))).toBeTrue();
    expect(tsAxis.contains(sec(150))).toBeFalse();
    expect(tsAxis.pointValue(valueMap(
               {key: 'x_axis', val: ts(sec(50))},
               )))
        .toEqual(sec(50));
    expect(() => {
      tsAxis.pointValue(valueMap(
          {key: 'missing field', val: ts(sec(50))},
          ));
    }).toThrow();
    expect(tsAxis.atOffset(d(50 * 1E9))).toEqual(sec(50));
    expect(tsAxis.toDomainFraction(sec(25))).toEqual(.25);
  });

  it('defines duration axes and gets points', () => {
    const axis = getAxis(valueMap(
        {key: 'category_defined_id', val: str('x_axis')},
        {key: 'category_display_name', val: str('time from start')},
        {key: 'category_description', val: str('Time from start')},
        {key: 'axis_type', val: str('duration')},
        {key: 'axis_min', val: dur(d(0))},
        {key: 'axis_max', val: dur(d(100))},
        ));
    expect(axis).toBeInstanceOf(DurationAxis);
    const dAxis = axis as DurationAxis;
    expect(dAxis.contains(d(50))).toBeTrue();
    expect(dAxis.contains(d(150))).toBeFalse();
    expect(dAxis.pointValue(valueMap(
               {key: 'x_axis', val: dur(d(50))},
               )))
        .toEqual(d(50));
    expect(() => {
      dAxis.pointValue(valueMap(
          {key: 'missing field', val: dur(d(50))},
          ));
    }).toThrow();
    expect(dAxis.toDomainFraction(d(25))).toEqual(.25);
  });

  it('defines number axes and gets points', () => {
    const axis = getAxis(valueMap(
        {key: 'category_defined_id', val: str('y_axis')},
        {key: 'category_display_name', val: str('events per second')},
        {key: 'category_description', val: str('Events per second')},
        {key: 'axis_type', val: str('double')},
        {key: 'axis_min', val: dbl(0)},
        {key: 'axis_max', val: dbl(100)},
        ));
    expect(axis).toBeInstanceOf(NumberAxis);
    const dblAxis = axis as NumberAxis;
    expect(dblAxis.contains(50)).toBeTrue();
    expect(dblAxis.contains(150)).toBeFalse();
    expect(dblAxis.pointValue(valueMap(
               {key: 'y_axis', val: dbl(50)},
               )))
        .toEqual(50);
    expect(() => {
      dblAxis.pointValue(valueMap(
          {key: 'missing field', val: dbl(50)},
          ));
    }).toThrow();
    expect(dblAxis.toDomainFraction(25)).toEqual(.25);
    expect(dblAxis.toDomainFraction(125)).toEqual(1.25);
  });

  it('unions axes properly', () => {
    const axis1 = getAxis(valueMap(
        {key: 'category_defined_id', val: str('x_axis')},
        {key: 'category_display_name', val: str('time from start')},
        {key: 'category_description', val: str('Time from start')},
        {key: 'axis_type', val: str('timestamp')},
        {key: 'axis_min', val: ts(sec(50))},
        {key: 'axis_max', val: ts(sec(100))},
        ));
    const axis2 = getAxis(valueMap(
        {key: 'category_defined_id', val: str('x_axis')},
        {key: 'category_display_name', val: str('time from start')},
        {key: 'category_description', val: str('Time from start')},
        {key: 'axis_type', val: str('timestamp')},
        {key: 'axis_min', val: ts(sec(0))},
        {key: 'axis_max', val: ts(sec(150))},
        ));
    const axis3 = getAxis(valueMap(
        {key: 'category_defined_id', val: str('x_axis')},
        {key: 'category_display_name', val: str('time from start')},
        {key: 'category_description', val: str('Time from start')},
        {key: 'axis_type', val: str('duration')},
        {key: 'axis_min', val: dur(d(0))},
        {key: 'axis_max', val: dur(d(100))},
        ));
    expect(() => unionAxes(axis1, axis2, axis3)).toThrow();
    const uAxis = unionAxes(axis1, axis2);
    expect(uAxis).toBeInstanceOf(TimestampAxis);
    const tsAxis = uAxis as TimestampAxis;
    expect(tsAxis.min).toEqual(sec(0));
    expect(tsAxis.max).toEqual(sec(150));
  });
});
